### ISSUE TITLE
[RFC] Component Overrides Standardization & Documentation

### ISSUE THREAD

**Issue by abmai** [uber-web/baseui#375]
### Current State of Overrides API

We have a fairly powerful overrides API. However, I think there are two issues we need to address sooner rather than later.

1. It is not clearly documented to the users and is not easily understood at each component's level. I have received a few repeated questions from users about how to override parts of their components.

2. There is a lack of standardization on how we approach overrides for components that re-use our other components. Do we expose a top-level override API for the entire component that's re-used, and expect users to make granular overrides by writing their own layer on top, or do we expose all the granular overrides of the re-used component?

### Documentation Issue & Proposal

Our `overrides` API could be used in two ways:

```js
const overrides = {
  Root: ReactComponent
};
// OR
const overrides = {
  Root: {
    component: ReactComponent,
    props: {...}
    style: {...}
       // OR
    style: ({$theme /* and other presentational props */}) => {...}
  }
};
```

This is **not** documented as a general API that **every** component follows. We do document this for every component but it's not obvious at first glance. There are also some test coverage concerns raised in the `style: ({$theme}) => {...}` approach (anon function declared inline).

Secondly, it's not obvious which component the user should override to achieve a particular behavior. In the case of `Pagination`, we have the following overrides
```
overrides: {PrevButton, NextButton, DropdownContainer, DropdownMenu, DropdownButton}
```
How can the user know the difference between `DropdownContainer` and `DropdownMenu` and decide which one to override without 1) reading the source code or 2) trial and error.

And in the case of the `Modal` component, which exposes the following overrides:
```
overrides: {Root, Backdrop, Dialog, DialogContainer, Close}
```
I'm not able to immediately understand the role of `Root`, `Dialog`, `DialogContainer` in relation to one another from this.

This extends to all of our other components that use the `overrides` API. A subset of the overrides are obvious, `Close`, `PrevButton`, etc. but many are not.

In order to fully utilize this overrides API, the user must have a clear understanding of the DOM structure and what each sub-component does.

#### Proposal

1. We need a high-level `overrides` API documentation with some examples of how it's used.
2. Standardize on some sub-components, like `Root` - i.e. always the outermost `<div>` of the component.
3. Have an `Overrides API` section for each component's RFC that describes the presentational responsibility of each sub-component.
4. It would be great to have a visual layered description of the component made up of its subcomponent. I don't know how to achieve this though.

### Components Re-use & Overrides

`Select`, for example, currently uses `Menu` underneath. There is no way to override the entire `Menu` component at the top level. `Pagination` also uses `Button`, but its API only allows overriding of the `BaseButton` sub-component inside `Button`. It's also gearing up to re-use `Select` entirely.

Should both of these components allow only top-level override of `Menu`, `Button`, and `Select`?  Or should, say, `Select` continue to expose more granular override options. But what if the user wants to override parts of `Menu` that's not exposed as an override by `Select`? 

#### Proposal

1. Allow only high-level override of re-used components at the top API. If the user wants to override the internals of those component, they can do so separately:
```js
function CustomMenu() {
  return (
    <Menu
      overrides={{
        Option: {...}
      }}
    />
  );
}

function MyComponent() {
  return (
    <Select
      overrides={{
        Menu: CustomMenu
      }}
     />
  );
}
```

2. Expose *all* overrides of the re-used component, and reconstruct the re-used component internally with those overrides. This would expand the override API of `Select` to include all the overrides offered by `Menu`, and expand all of `Pagination` overrides to include all the overrides offered by `Select` (which includes all the overrides of `Menu` and so on...).

### Conclusion

I don't have a clear answer to any of these. But I figured that with our components count increasing, it would be good to align on these issues, before we start building more complex components.

Thoughts?

**Comment by schnerd** [comment]
Good timing – #380 is an example of a user getting blocked by **Components Re-use & Overrides**

**Comment by blackswanny** [comment]
> Select, for example, currently uses Menu underneath. There is no way to override the entire Menu component at the top level. 

We do not allow to entirely override functional components as it will literally not work and possible break it. I think the main idea of overrides is for granular changes, mostly styles.

> It would be great to have a visual layered description of the component made up of its subcomponent. I don't know how to achieve this though.

This is the great way to explain subcomponents. Something like this:
![image](https://user-images.githubusercontent.com/1851469/46887910-d6e78e00-ce13-11e8-9a16-1ab70e5c6e57.png)


And we need to decide whether to document all overrides in the very top component and pass them through, or still explain that user can use overrides of re-used sub component if needed. I prefer fisrt approach as it clears out any guesses, no need to go to another component's API and check how it works. Interface is plain and gathered under one component's API

**Comment by schnerd** [comment]
@blackswanny can you clarify what you're advocating for by showing a code example for how Select/Menu overrides would work?

**Comment by blackswanny** [comment]
@schnerd We can do
```js
    <Select
      overrides={{
        Menu: {
             props: {
                  overrides: {MenuItem}
             }
        }
      }}
     />
```
or
```js
    <Select
      overrides={{
        MenuItem
      }}
     />
```

I prefer second option. Only very most parent declares overrides , which are passed to all internal. For that all overrides should be unique across all Base UI . So instead of `Root` we would have `SelectRoot`, `MenuRoot`, etc

**Comment by schnerd** [comment]
> For that all overrides should be unique across all Base UI . So instead of Root we would have SelectRoot, MenuRoot, etc

I don't think the name exposed by parent component needs **exactly** match the one exposed by the it's child. Providing a different name could actually be more descriptive in the public API. 

This is especially useful if a component is reused twice. Datepicker will have `Button`s to switch between months and `Button`s for quick select ('3 months', etc). These buttons look and function differently, you'd probably want to override them separately. So rather than exposing a single `BaseButton` override in Datepicker, you might want `QuickSelectBaseButton` (or `QuickSelectButton: {BaseButton: {}}`) as well as `MonthPagerBaseButton`.

Datepicker is actually an interesting example since it forces us to think about how well each solution scales. If Datepicker needs to manually expose all its elements and its subcomponents elements, it could _easily_ end up with 30 or so top-level overrides:

```js
<Datepicker overrides={{
  DateInputContainer
  DateInputInput
  DateInputBefore
  DateInputAfter
  DateInputArrow
  PopoverBody
  DateRangeHeader
  MonthPaginateButtonContainer
  MonthPaginateButtonBase
  MonthPaginateButtonStartEnhancer
  MonthPaginateButtonEndEnhancer
  YearMonthSelector
  YearMonthSelect
  YearMonthSelectInput
  YearMonthSelectSearchIcon
  YearMonthSelectMenu
  YearMonthSelectMenuList
  YearMonthSelectMenuOption
  CalendarContainer
  CalendarMonth
  CalendarDayOfWeekHeader
  CalendarDayOfWeek
  CalendarWeekRow
  CalendarDay
  QuickSelectTitle
  QuickSelectContainer
  QuickSelectButtonContainer
  QuickSelectButtonBase
  QuickSelectButtonStartEnhancer
  QuickSelectButtonEndEnhancer
}}>
```

This create a really complicated public API, and then we also have the extra burden of propagating 
and mapping each of these keys down to the sub-component where it needs to be used.

```js
class DatePicker {
  render() {
    const [Root, rootProps] = getOverrides(overrides.Root, StyledRoot);
    
    // Destructure 30-40 overrides here so we can pass down to appropriate subcomponents
    const DateInputOverrides = {
      DateInputContainer: overrides.DateInputContainer,
      DateInputInput: overrides.DateInputInput,
      DateInputBefore: overrides.DateInputBefore,
      DateInputAfter: overrides.DateInputAfter,
      DateInputArrow: overrides.DateInputArrow
    };
    const DatePopoverOverrides = {
      PopoverBody: PopoverBody,
      DateRangeHeader
      MonthPaginateButtonContainer
      MonthPaginateButtonBase
      MonthPaginateButtonStartEnhancer
      MonthPaginateButtonEndEnhancer
      YearMonthSelector
      YearMonthSelect
      YearMonthSelectInput
      YearMonthSelectSearchIcon
      YearMonthSelectMenu
      YearMonthSelectMenuList
      YearMonthSelectMenuOption
      // etc.
    };
    return (
      <Root>
        <DateInputs overrides={DateInputOverrides}>
        {this.state.isFocused && (
          <DatePopover overrides={DatePopoverOverrides} />
        )}
      </Root>
    );
  }
}

class DatePopover {
  render() {
    // Similar large mappings down to DateRangeHeader, DateCalendars, QuickSelect, etc.
  }
}
```

If we allow overrides to be more hierarchical, the public API might look like the following:

```js
<Datepicker overrides={{
  DateInputs: {
    Container,
    Arrow
    Input: {InputContainer, Input, Before, After}
  },
  DatePopover: {
    PopoverBody,
    DateRangeHeader: {
      MonthPaginateButton: {
        Container,
        Button: {BaseButton, StartEnhancer, EndEnhancer}
      },
      YearMonthSelector: {
        Select: {
          Input, 
          SearchIcon,
          Menu: {List, Option}
        }
      }
    },
    Calendars: {
      Container,
      Month,
      DayOfWeekHeader,
      DayOfWeek,
      Day,
    }
    QuickSelect: {
      Container,
      Title,
      Button: {BaseButton, StartEnhancer, EndEnhancer}
    },
  }
}}>
```

Arguable whether that's better or worse than the flattened list. But it seems like the real benefit here is how much it simplifies the internal implementation:

```js
function DatePicker({overrides}) {
  const [Root, rootProps] = getOverrides(overrides.Root, StyledRoot);
  return (
    <Root>
      <DateInputs overrides={overrides.DateInputs}>
      {this.state.isFocused && (
        <DatePopover overrides={overrides.DatePopover} />
      )}
    </Root>
  );
}
```

Each subcomponent is just responsible for passing along an object down to its child, all the way down to something like `Button` in `QuickSelect`.

One other benefit of this – if someone decides to add another override element in `Button` all components using Button will get it for free, no need to update their source code and manually pass down yet another override.

**Comment by blackswanny** [comment]
@schnerd yes, it can be unique only inside of exact component. So if we had original `Root` in `Menu`, it should be renamed and exposed under different name in `Select` usage.
I like the hierarchy you provided here. As it's still under one `overrides` property, which belongs to the very top component. But I would like not to use example I shown above, where we dig into internal props of innner component and provide `overrides` property very deeply nested there

**Comment by nadiia** [comment]
Thank you @abmai for bringing it up and creating the issue and proposals for it! This is a growing issue of inconsistency and confusion. We already have in place the different approaches mentioned here and we do need to standardize on it and make sure we expose overrides for the reused components in a consistent way across the library.
I'm in favor of the 3rd option David mentioned:
```
 <Pagination
  overrides={{
    Root: {props: {}, style: {}}
    Select: React.Component | {
      Menu: React.Component | {
        Option: React.Component | {style: {color: 'green'}}
      }
    }
  }}
/>
```
One more thing that I'd like to bring up here is whether or not we're going to allow passing any top-level props to the re-used components. The top-level props of any re-used component could be handled in a custom component override case: 
```
function CustomSelect() {
  return (
    <Select
      getSelectedOptionLabel: (option) => React.Node,
      overrides={{
        Menu: {...}
      }}
    />
  );
}

function MyComponent() {
  return (
    <Pagination
      overrides={{
        Select: CustomSelect
      }}
     />
  );
}
```
But that may not work well for any dynamic prop.
Should we include those props in the 3rd proposal option of the "Components Re-use & Overrides"? Could be something like:
```
<Pagination
  overrides={{
    Root: {props: {}, style: {}},
    Select: React.Component | {
      ...{getSelectedOptionLabel: (option) => React.Node}, 
      overrides: {
        Menu: React.Component | {
          overrides: {
            Option: React.Component | {style: {color: 'green'}}
          }
        }
      }
    }}
  }}
/>
```

**Comment by schnerd** [comment]
@nadiia great question. I'm inclined to say that we should just keep it simple for now and not allow props overrides for reused components internally, since that helps keep the overrides hierarchy a bit less hectic and people can still do the component replacement if they need to. We could also selectively expose certain props in the top-level component (like offering getSelectPageLabel and passing that down to getSelectedOptionLabel). That's my hunch, I could be swayed though if this seems like a very common use case.

**Comment by tlrobinson** [comment]
Just wanted to chime in with my "overrides" implementation using hooks or higher-order components to eliminates all the manual prop spreading: https://github.com/tlrobinson/overrides

(caveat: I just hacked this together last night and haven't used it for anything real yet. It's possible I'm missing a reason why this is a terrible idea. The only downside I've come up with is the extra layer of wrapper components it introduces)

> Arguable whether that's better or worse than the flattened list. But it seems like the real benefit here is how much it simplifies the internal implementation:

FWIW I've implemented nested hooks without flattening. It's very simple to implement and I think it's better to be explicit. And `overrides={{ Calender: { Month: ... } }}` isn't much more verbose than `overrides={{ CalenderMonth: ... }}`, for example.
